File: internal/report/report.go
// Package report renders comparison results as human tables or JSON.
package report

import (
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/newrelic-forks/otel-metric-validator/internal/compare"
	"github.com/newrelic-forks/otel-metric-validator/internal/ingestcheck"
)

var dbOrder = map[string]int{
	compare.Mismatch:        0,
	compare.MissingInIngest: 1,
	compare.OK:              2,
	compare.MissingInDB:     3,
	compare.Skipped:         4,
}

var ingestOrder = map[string]int{
	ingestcheck.Mismatch: 0,
	ingestcheck.Error:    1,
	ingestcheck.NoData:   2,
	ingestcheck.OK:       3,
	ingestcheck.Skipped:  4,
}

func attrsStr(attrs map[string]string) string {
	if len(attrs) == 0 {
		return ""
	}
	parts := make([]string, 0, len(attrs))
	for k, v := range attrs {
		parts = append(parts, k+"="+v)
	}
	sort.Strings(parts)
	return strings.Join(parts, ",")
}

func fmtNum(v *float64) string {
	if v == nil {
		return "-"
	}
	if *v == math.Trunc(*v) && !math.IsInf(*v, 0) {
		return fmt.Sprintf("%d", int64(*v))
	}
	return fmt.Sprintf("%.4g", *v)
}

// delta renders rel_diff as a percentage when expected is non-zero, else absolute.
func delta(rel, expected *float64) string {
	if rel == nil {
		return ""
	}
	if expected != nil && *expected != 0 {
		return fmt.Sprintf("%.2f%%", *rel*100)
	}
	return fmt.Sprintf("%.4g", *rel)
}

func table(headers []string, lines [][]string) string {
	widths := make([]int, len(headers))
	for i, h := range headers {
		widths[i] = len(h)
	}
	for _, ln := range lines {
		for i, cell := range ln {
			if len(cell) > widths[i] {
				widths[i] = len(cell)
			}
		}
	}
	fmtRow := func(cells []string) string {
		padded := make([]string, len(cells))
		for i, c := range cells {
			padded[i] = c + strings.Repeat(" ", widths[i]-len(c))
		}
		return strings.Join(padded, "  ")
	}
	sep := make([]string, len(headers))
	for i, w := range widths {
		sep[i] = strings.Repeat("-", w)
	}
	out := []string{fmtRow(headers), fmtRow(sep)}
	for _, ln := range lines {
		out = append(out, fmtRow(ln))
	}
	return strings.Join(out, "\n")
}

// RenderTable renders the DB-check results.
func RenderTable(results []compare.Result, showOK bool) string {
	rows := make([]compare.Result, 0, len(results))
	for _, r := range results {
		if showOK || r.Status != compare.OK {
			rows = append(rows, r)
		}
	}
	sort.SliceStable(rows, func(i, j int) bool {
		oi, oj := order(dbOrder, rows[i].Status), order(dbOrder, rows[j].Status)
		if oi != oj {
			return oi < oj
		}
		return rows[i].Metric < rows[j].Metric
	})
	var lines [][]string
	for _, r := range rows {
		lines = append(lines, []string{
			r.Status, r.Metric, attrsStr(r.Attrs),
			fmtNum(r.Expected), fmtNum(r.Actual), delta(r.RelDiff, r.Expected),
		})
	}
	return table([]string{"STATUS", "METRIC", "ATTRS", "EXPECTED(DB)", "ACTUAL(OTEL)", "Δ"}, lines)
}

// RenderIngestTable renders the ingest-check results.
func RenderIngestTable(results []ingestcheck.Result, showOK bool) string {
	rows := make([]ingestcheck.Result, 0, len(results))
	for _, r := range results {
		if showOK || r.Status != ingestcheck.OK {
			rows = append(rows, r)
		}
	}
	sort.SliceStable(rows, func(i, j int) bool {
		oi, oj := order(ingestOrder, rows[i].Status), order(ingestOrder, rows[j].Status)
		if oi != oj {
			return oi < oj
		}
		return rows[i].Metric < rows[j].Metric
	})
	var lines [][]string
	for _, r := range rows {
		kind := ""
		if r.ValueType == "sum" {
			kind = "Δ"
		} else if r.ValueType != "" {
			kind = "latest"
		}
		lines = append(lines, []string{
			r.Status, r.Metric, attrsStr(r.Attrs), kind,
			fmtNum(r.Expected), fmtNum(r.Actual), delta(r.RelDiff, r.Expected),
		})
	}
	return table([]string{"STATUS", "METRIC", "ATTRS", "TYPE", "EXPECTED", "NRDB", "Δ"}, lines)
}

// RenderJSON renders the DB-check results as indented JSON.
func RenderJSON(results []compare.Result) string { return mustJSON(results) }

// RenderIngestJSON renders the ingest-check results as indented JSON.
func RenderIngestJSON(results []ingestcheck.Result) string { return mustJSON(results) }

// SummaryLine renders status counts as "K=v  K2=v2".
func SummaryLine(counts map[string]int) string {
	if len(counts) == 0 {
		return "(no results)"
	}
	keys := make([]string, 0, len(counts))
	for k := range counts {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, len(keys))
	for i, k := range keys {
		parts[i] = fmt.Sprintf("%s=%d", k, counts[k])
	}
	return strings.Join(parts, "  ")
}

func order(m map[string]int, status string) int {
	if v, ok := m[status]; ok {
		return v
	}
	return 99
}

func mustJSON(v any) string {
	b, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return "[]"
	}
	return string(b)
}

File: validator/report.py
"""Render comparison results as a human table or JSON."""

from __future__ import annotations

import json

from .comparator import Result
from .comparator_status import MISMATCH, MISSING_IN_DB, MISSING_IN_INGEST, OK, SKIPPED

_ORDER = [MISMATCH, MISSING_IN_INGEST, OK, MISSING_IN_DB, SKIPPED]


def _attrs_str(attrs: dict) -> str:
    if not attrs:
        return ""
    return ",".join(f"{k}={v}" for k, v in sorted(attrs.items()))


def _fmt(v: float | None) -> str:
    if v is None:
        return "-"
    if v == int(v):
        return str(int(v))
    return f"{v:.4g}"


def render_table(results: list[Result], show_ok: bool = True) -> str:
    rows = [r for r in results if show_ok or r.status != OK]
    rows.sort(key=lambda r: (_ORDER.index(r.status) if r.status in _ORDER else 99, r.metric))

    headers = ("STATUS", "METRIC", "ATTRS", "EXPECTED(DB)", "ACTUAL(OTEL)", "Δ")
    lines = []
    for r in rows:
        delta = "" if r.rel_diff is None else (f"{r.rel_diff*100:.2f}%" if r.expected else f"{r.rel_diff:.4g}")
        lines.append((
            r.status, r.metric, _attrs_str(r.attrs),
            _fmt(r.expected), _fmt(r.actual), delta,
        ))

    widths = [len(h) for h in headers]
    for ln in lines:
        for i, cell in enumerate(ln):
            widths[i] = max(widths[i], len(cell))

    def fmt_row(cells):
        return "  ".join(c.ljust(widths[i]) for i, c in enumerate(cells))

    out = [fmt_row(headers), fmt_row(["-" * w for w in widths])]
    out += [fmt_row(ln) for ln in lines]
    return "\n".join(out)


_INGEST_ORDER = ["INGEST_MISMATCH", "INGEST_ERROR", "INGEST_NO_DATA", "INGEST_OK", "INGEST_SKIPPED"]


def render_ingest_table(results, show_ok: bool = True) -> str:
    rows = [r for r in results if show_ok or r.status != "INGEST_OK"]
    rows.sort(key=lambda r: (_INGEST_ORDER.index(r.status) if r.status in _INGEST_ORDER else 99, r.metric))

    headers = ("STATUS", "METRIC", "ATTRS", "TYPE", "EXPECTED", "NRDB", "Δ")
    lines = []
    for r in rows:
        kind = "Δ" if r.value_type == "sum" else ("latest" if r.value_type else "")
        delta = "" if r.rel_diff is None else (f"{r.rel_diff*100:.2f}%" if r.expected else f"{r.rel_diff:.4g}")
        lines.append((r.status, r.metric, _attrs_str(r.attrs), kind,
                      _fmt(r.expected), _fmt(r.actual), delta))

    widths = [len(h) for h in headers]
    for ln in lines:
        for i, cell in enumerate(ln):
            widths[i] = max(widths[i], len(cell))

    def fmt_row(cells):
        return "  ".join(c.ljust(widths[i]) for i, c in enumerate(cells))

    out = [fmt_row(headers), fmt_row(["-" * w for w in widths])]
    out += [fmt_row(ln) for ln in lines]
    return "\n".join(out)


def render_ingest_json(results) -> str:
    return json.dumps(
        [
            {
                "status": r.status, "metric": r.metric, "attrs": r.attrs,
                "expected": r.expected, "actual": r.actual, "rel_diff": r.rel_diff,
                "value_type": r.value_type, "nrql": r.nrql, "note": r.note,
            }
            for r in results
        ],
        indent=2,
    )


def render_json(results: list[Result]) -> str:
    return json.dumps(
        [
            {
                "status": r.status,
                "metric": r.metric,
                "attrs": r.attrs,
                "expected": r.expected,
                "actual": r.actual,
                "rel_diff": r.rel_diff,
                "value_type": r.value_type,
                "note": r.note,
            }
            for r in results
        ],
        indent=2,
    )


def summary_line(counts: dict[str, int]) -> str:
    parts = [f"{k}={v}" for k, v in sorted(counts.items())]
    return "  ".join(parts) if parts else "(no results)"
